File: python/armstrong-numbers/armstrong_numbers.py
from typing import List


def _get_digits(number: int) -> List[int]:
    digits = []
    while number >= 10:
        digits.append(number % 10)
        number //= 10
    digits.append(number)
    return digits


def is_armstrong_number(number: int) -> bool:
    digits = _get_digits(number)
    print(digits)
    return sum(dig ** len(digits) for dig in digits) == number

File: c/armstrong-numbers/armstrong_numbers.c
#include <math.h>
#include <stdint.h>
#include <stdlib.h>
#include "armstrong_numbers.h"

#define NDIGITS 20

/* This is to silence -Werror=missing-declarations while allowing it to remain
 * private / not in the header. */
size_t extract_digits(int number, uint8_t* digits);

size_t extract_digits(int number, uint8_t* digits) {
    size_t offset = 0;
    while (number >= 10) {
        digits[offset] = number % 10;
        number /= 10;
        offset += 1;
    }
    digits[offset] = number;
    return offset + 1;
}

bool is_armstrong_number(int candidate) {
    uint8_t digits[NDIGITS];
    const size_t ndigits = extract_digits(candidate, &digits[0]);
    int acc = 0;
    for (size_t i = 0; i < ndigits; i++) {
        acc += pow(digits[i], ndigits);
    }
    return candidate == acc;
}
